exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java

```diff
@@ -226,6 +244,96 @@ public IterOutcome next() {
+    if (runtimeFilterWritable == null) {
+      return;
+    }
+    if (recordCount <= 0) {
+      return;
+    }
+    List<BloomFilter> bloomFilters = runtimeFilterWritable.unwrap();
+    if (hash64 == null) {
+      ValueVectorHashHelper hashHelper = new ValueVectorHashHelper(this, context);
+      try {
+        //generate hash helper
+        this.toFilterFields = runtimeFilterWritable.getRuntimeFilterBDef().getProbeFieldsList();
+        List<LogicalExpression> hashFieldExps = new ArrayList<>();
+        List<TypedFieldId> typedFieldIds = new ArrayList<>();
+        for (String toFilterField : toFilterFields) {
+          SchemaPath schemaPath = new SchemaPath(new PathSegment.NameSegment(toFilterField), ExpressionPosition.UNKNOWN);
+          TypedFieldId typedFieldId = container.getValueVectorId(schemaPath);
+          this.field2id.put(toFilterField, typedFieldId.getFieldIds()[0]);
+          typedFieldIds.add(typedFieldId);
+          ValueVectorReadExpression toHashFieldExp = new ValueVectorReadExpression(typedFieldId);
+          hashFieldExps.add(toHashFieldExp);
+        }
+        hash64 = hashHelper.getHash64(hashFieldExps.toArray(new LogicalExpression[hashFieldExps.size()]), typedFieldIds.toArray(new TypedFieldId[typedFieldIds.size()]));
+      } catch (Exception e) {
+        throw UserException.internalError(e).build(logger);
+      }
+    }
+    selectionVector2.allocateNew(recordCount);
+    BitSet bitSet = new BitSet(recordCount);
+    for (int i = 0; i < toFilterFields.size(); i++) {
+      BloomFilter bloomFilter = bloomFilters.get(i);
+      String fieldName = toFilterFields.get(i);
+      computeBitSet(field2id.get(fieldName), bloomFilter, bitSet);
+    }
+    int svIndex = 0;
+    int tmpFilterRows = 0;
+    for (int i = 0; i < recordCount; i++) {
+      boolean contain = bitSet.get(i);
+      if (contain) {
+        selectionVector2.setIndex(svIndex, i);
+        svIndex++;
+      } else {
+        tmpFilterRows++;
+      }
+    }
+    selectionVector2.setRecordCount(svIndex);
+    if (tmpFilterRows > 0 && tmpFilterRows == recordCount) {
+      recordCount = 0;
+      selectionVector2.clear();
+      logger.debug("filter {} rows by the RuntimeFilter", tmpFilterRows);
+      return;
+    }
+    if (tmpFilterRows > 0 && tmpFilterRows != recordCount ) {
+      totalFilterRows = totalFilterRows + tmpFilterRows;
+      recordCount = svIndex;
+      BatchSchema batchSchema = this.schema;
+      VectorContainer backUpContainer = new VectorContainer(this.oContext.getAllocator(), batchSchema);
+      int fieldCount = batchSchema.getFieldCount();
+      for (int i = 0; i < fieldCount; i++) {
+        ValueVector from = this.getContainer().getValueVector(i).getValueVector();
+        ValueVector to = backUpContainer.getValueVector(i).getValueVector();
+        to.setInitialCapacity(svIndex);
+        for (int r = 0; r < svIndex; r++) {
+          to.copyEntry(r, from, selectionVector2.getIndex(r));
```

Comment:
To summarize the way you have implemented this part:  suppose the original ScanBatch contained 100 rows and 10 of them qualified the bloom filter, you create an SV2 of size 10, set the original qualifying row's index in the SV2, then for each ValueVector you copy the qualifying row's data into the backupContainer, followed by exchanging it with the current output container.   A couple of thoughts about this: 
  - Why not produce the SV2 in the output batch and let the downstream operator handle it ? Quite often there may be a Filter operator above the Scan  which would be applying other filters (i.e not the run-time filters) and it can combine its own SV2 with the SV2 produced by the Scan.  That way you avoid the extra copying and let Filter handle it (note that Filter does code-gen so it is more efficient way to handle bulk filtering). 
 - There is clearly trade-offs with the extra copy approach : it depends on selectivity, i.e how many rows actually get eliminated by the run-time filter.  I suppose you have mentioned this as a TODO depending on the NDV statistics ?

Reply:
@aman thanks for other valuable reviews ,will soon update that. To this point, I initially tend to output the SV2. But to fellow reasons, I give up: 

* Not all the possible operators support the SelectionModel. If users’ rule pushed down the filter conditions,the output SV2 maybe not processed by the not supported operators(i.e. it’s not definitive  to have a operator which supports the SelectionModel above the Scan node).

* The BatchSchema’s SelectionModel also becomes a runtime var.This will also affect the upper filter node’s code-gen logic to dynamically generate fresh filter codes to the Scan’s SV2.

I agree that the memory copy cost will be less if the above filter node can filter more rows over the Scan’s SV2.So what your opinion about this?

Reply:
Hi @weijietong ,  the physical plan operators are supposed to implement 2 APIs:  `getSupportedEncodings()` API (see [1]) which indicates whether or not it accepts an SelectionVector2 (other options are SV4 and NONE) and the corresponding `getEncoding()` API.  If an operator does not accept an SV2 or SV4, the planner will insert a SelectionVectorRemover just below that node which essentially does the copying of qualified rows.  Note that the SVRemover is implemented by RemovingRecordBatch [2] which uses a `StraightCopier` whenever the child has NONE, so in that case it just does a simple transfer instead of copy.  

In the case of Filter,  it can accept both NONE and SV2 because it is possible in some cases to have a filter on top of another filter (with some intermediate non-blocking operator).  

Let me know if this makes sense.   The main reason I am proposing this is the Filter-Scan is a very common pattern and we would want to minimize the copy overhead as much as possible.  Unfortunately, I am tied up with some other work otherwise I could make the changes on top of your branch to experiment. 

[1] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/Prel.java#L51
[2] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/RemovingRecordBatch.java#L57